### PR TITLE
Support multiple SI campaigns in a day

### PR DESCRIPTION
**Pull request by simonhong**
NTPSponsoredImagesData manages `Campaign` list instead of background list.
And each `Campaigin` represents each campaign that includes its own background list and logo.

This is backward-compatible.
If photo.json doesn't have campaigns list, global campaign will be the only item of campaigns list.

Resolves https://github.com/brave/brave-browser/issues/19999

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

`npm run test brave_unit_tests -- --filter=*NTPBackground*`
`npm run test brave_unit_tests -- --filter=*ViewCounter*`

### CODE REVIEW

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc [hunk 1]
```diff
@@ -145,25 +148,70 @@ NTPSponsoredImagesData::NTPSponsoredImagesData(
...
+    url_prefix += kSuperReferralPath;
+  } else {
+    url_prefix += kSponsoredImagesPath;
+  }
+
+  auto* campaigns_value = json_value->FindListKey(kCampaignsKey);
+  if (campaigns_value) {
+    ParseCampaignsList(campaigns_value->Clone(), installed_dir);
+  } else {
+    // Get global campaign directly if campaigns list isn't existed.
```
Comment (wchen342) [inline components/ntp_background_images/browser/ntp_sponsored_images_data.cc:164]: Typo?

File: components/ntp_background_images/browser/view_counter_model.cc [hunk 2]
```diff
@@ -16,6 +17,23 @@ ViewCounterModel::ViewCounterModel() {
...
 ViewCounterModel::~ViewCounterModel() = default;
 
+void ViewCounterModel::SetCampaignsTotalBrandedImageCount(
+    const std::vector<size_t>& campaigns_total_image_count) {
+  campaigns_total_branded_image_count_ = campaigns_total_image_count;
+  total_campaign_count_ = campaigns_total_branded_image_count_.size();
+  for (size_t i = 0; i < total_campaign_count_; ++i) {
+    campaigns_current_branded_image_index_.push_back(0);
+  }
+  // Pick first campaign randonly.
```
Comment (wchen342) [inline components/ntp_background_images/browser/view_counter_model.cc:27]: Typo

**Review by wchen342 — APPROVED** [review]
LGTM:+1:

File: components/ntp_background_images/browser/view_counter_model.cc (hunk 2 above)
Comment (simonhong) [reply components/ntp_background_images/browser/view_counter_model.cc]: Updated to `Pick the first campaign randomly.`

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc (hunk 1 above)
Comment (simonhong) [reply components/ntp_background_images/browser/ntp_sponsored_images_data.cc]: Updated to `Get a global campaign directly if the campaign list doesn't exist.`.

File: components/ntp_background_images/browser/view_counter_service.cc [hunk 3]
```diff
@@ -76,10 +76,11 @@ ViewCounterService::ViewCounterService(NTPBackgroundImagesService* service,
   branded_new_tab_count_state_ =
       std::make_unique<WeeklyStorage>(local_state, kSponsoredNewTabsCreated);
 
-  if (auto* data = GetCurrentBrandedWallpaperData())
-    model_.set_total_branded_image_count(data->backgrounds.size());
-  if (auto* data = GetCurrentWallpaperData())
-    model_.set_total_image_count(data->backgrounds.size());
+  ResetModel();
+  // if (auto* data = GetCurrentBrandedWallpaperData())
```
Comment (petemill) [inline components/ntp_background_images/browser/view_counter_service.cc:80]: did you mean to keep this commented-out code in?

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc [hunk 4]
```diff
@@ -145,25 +148,70 @@ NTPSponsoredImagesData::NTPSponsoredImagesData(
...
     theme_name = *name;
-    DVLOG(2) << __func__ << ": Theme name: " << theme_name;
+    url_prefix += kSuperReferralPath;
+  } else {
+    url_prefix += kSponsoredImagesPath;
+  }
+
+  auto* campaigns_value = json_value->FindListKey(kCampaignsKey);
+  if (campaigns_value) {
+    ParseCampaignsList(campaigns_value->Clone(), installed_dir);
```
Comment (petemill) [inline components/ntp_background_images/browser/ntp_sponsored_images_data.cc:162]: Can't we pass as reference without having to clone the data?

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc [hunk 5]
```diff
@@ -145,25 +148,70 @@ NTPSponsoredImagesData::NTPSponsoredImagesData(
...
+  } else {
+    // Get a global campaign directly if the campaign list doesn't exist.
+    const auto campaign =
+        GetCampaignFromValue(json_value->Clone(), installed_dir);
+    if (campaign.IsValid())
+      campaigns.push_back(campaign);
+  }
```
Comment (petemill) [inline components/ntp_background_images/browser/ntp_sponsored_images_data.cc:169]: nit: how about a `VLOG` with the result of the parsing? As this is from external data, might help debug issues. Or do we have that somewhere else?

File: components/ntp_background_images/browser/view_counter_service.h [hunk 6]
```diff
@@ -63,8 +63,7 @@ class ViewCounterService : public KeyedService,
   base::Value GetCurrentWallpaperForDisplay() const;
   base::Value GetCurrentWallpaper() const;
   base::Value GetCurrentBrandedWallpaper() const;
-  std::vector<TopSite> GetTopSitesVectorForWebUI() const;
-  std::vector<TopSite> GetTopSitesVectorData() const;
+  std::vector<TopSite> GetTopSitesData() const;
```
Comment (petemill) [inline components/ntp_background_images/browser/view_counter_service.h:66]: 👍

**Review by petemill — COMMENTED** [review]
👍

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc (hunk 4 above)
Comment (simonhong) [reply components/ntp_background_images/browser/ntp_sponsored_images_data.cc:162]: Yup, can do. Done by https://github.com/brave/brave-core/pull/11628

File: components/ntp_background_images/browser/ntp_sponsored_images_data.cc (hunk 5 above)
Comment (simonhong) [reply components/ntp_background_images/browser/ntp_sponsored_images_data.cc:169]: Added log by https://github.com/brave/brave-core/pull/11628

File: components/ntp_background_images/browser/view_counter_service.cc (hunk 3 above)
Comment (simonhong) [reply components/ntp_background_images/browser/view_counter_service.cc:80]: It should be deleted :)
Deleted by https://github.com/brave/brave-core/pull/11628

**Pull request merged**